rir/src/interpreter/interp.cpp

```diff
@@ -2705,20 +2705,30 @@ SEXP rirExpr(SEXP f) {
     return f;
 }
 
-SEXP rirEval_f(SEXP f, SEXP env) {
-    assert(TYPEOF(f) == EXTERNALSXP);
-    Function* ff;
+SEXP rirEval_f(SEXP what, SEXP env) {
+    assert(TYPEOF(what) == EXTERNALSXP);
+    Code* c;
+    Function* f;
     DispatchTable* t;
     // TODO we do not really need the arg counts now
-    if (isValidCodeObject(f)) {
-        Code* c = (Code*)f;
-        return evalRirCode(c, globalContext(), env, 0);
-    } else if ((ff = Function::check(f))) {
-        return evalRirCode(ff->body(), globalContext(), env, 0);
-    } else if ((t = DispatchTable::check(f))) {
-        // Default target is the first version in the dispatch table
-        return evalRirCode(t->first()->body(), globalContext(), env, 0);
+    if ((c = isValidCodeObject(what))) {
+        // nothing
+    } else if ((f = Function::check(what))) {
+        // note: could try to optimize, but no way to update the
+        // containing dispatch table
+        // also this branch is probably not taken at all,
```

Comment:
Can we confirm if this branch is ever taken? If we're sure that it isn't, maybe we should just remove it?

I think this was originally my fault, when I added dispatch tables. Probably I didn't want to break or change too many things at once, and then never actually removed this branch.

Reply:
This is just my guess, not sure how to verify it. The reasoning is:

The function is not called anywhere in RIR, it is only registered in GNUR as an evaluation callback for EXTERNALSXPs. So the question is, can it happen that there is a naked Function object flying around somewhere in the open? The answer is hopefully no...

As for the Code branch, I believe that is used when promises are forced, so that one we need.

Reply:
Yeah, I agree that Code branch is needed, and Function branch seems unlikely.

If we expect all function objects to be in dispatch tables, then naked Function objects should be wrong... so maybe we should let this case fall to the assert?